### PR TITLE
add the selector text used when selector does not match element (closes #2568)

### LINKED ISSUE #2568: Show the selector text used when "The specified selector does not match any element in the DOM tree."
### Are you requesting a feature or reporting a bug?

New feature

### What is the current behavior?

screen shot 2018-06-27 at 10 35 44 am

When a selector is used that does not find anything, we print the error message 

```
The specified selector does not match any element in the DOM tree.
```

If using a page model like the above screenshot then the source code doesn't show the selector that was used.

### What is the expected behavior?

```
The specified selector does not match any element in the DOM tree:
[data-testid="navigation"] [data-testid="visible-items"] > a:nth-child(3) foobar
```

### How would you reproduce the current behavior (if this is a bug)?

* Use a selector that doesn't find anything on the page.

#### Provide the test code and the tested page URL (if applicable)

Tested page URL:

Any page

Test code

```js
const selector = '[data-testid="navigation"] [data-testid="visible-items"] > a:nth-child(3) foobar';
await t.click(Selector(selector));
```

### Specify your

* operating system: `macos`
* testcafe version: `0.20.4`
* node.js version: `10.5.0`

### PR DESCRIPTION
**Pull request by AlexKamaev**

### CODE REVIEW

File: src/errors/test-run/templates.js [hunk 1]
```diff
@@ -16,6 +17,30 @@ const SUBTITLES = {
     [TEST_RUN_PHASE.inBookmarkRestore]:       '<span class="subtitle">Error while restoring configuration after Role switch</span>\n'
 };
 
+function prepareSelectorCallstack (apiFnChain, apiFnIndex) {
+    if (typeof apiFnIndex === 'undefined')
+        return '';
+
+    const emptySpaces    = 10;
+    const elipsis        = '...)';
```
Comment (miherlosev) [inline src/errors/test-run/templates.js]: 2 `l` characters: `const ellipsis = ...`

File: test/functional/fixtures/regression/gh-2568/test.js [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+                    Cannot obtain information about the node because the specified selector does not match any node in the DOM tree.
+                      > | Selector('ul li')
+                `);
+            });
+    });
+
+    it('long selector', function () {
+        return runTests('testcafe-fixtures/index.js', 'long selector', { selectorTimeout: 100, shouldFail: true })
+            .catch(function (errs) {
+                expect(/Selector\('a+\.\.\.\)/.test(errs[0])).to.be.true;
```
Comment (miherlosev) [inline test/functional/fixtures/regression/gh-2568/test.js]: This test covers only ellipsis. You can convert this functional test to unit.

File: src/errors/test-run/templates.js [hunk 3]
```diff
@@ -16,6 +17,30 @@ const SUBTITLES = {
...
+    const viewportWidth  = getViewportWidth(process.stdout);
+    const availableWidth = viewportWidth - emptySpaces;
+
+    return apiFnChain.map((apiFn, index) => {
+        let formattedApiFn = String.fromCharCode(160);
```
[the remark this answers is not in the document]
Comment (AlexKamaev) [reply src/errors/test-run/templates.js]: we need &nbsp; here

[the remark this answers is not in the document]
Comment (AlexKamaev) [reply src/errors/test-run/templates.js]: we need `&nbsp;` here

File: src/client-functions/selectors/add-api.js [hunk 4]
```diff
@@ -310,11 +311,36 @@ function ensureRegExpContext (str) {
...
+        });
+        args = args.join(', ');
+
+        return `.${fn}(${args})`;
+    };
+
+    return function () {
+        apiFn = prepareApiFnArgs(arguments);
+
+        return originalFn.apply(0, arguments);
```
[the remark this answers is not in the document]
Comment (miherlosev) [reply src/client-functions/selectors/add-api.js]: So, then fix it and add a test if it's possible.

File: src/errors/test-run/templates.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexKamaev) [reply src/errors/test-run/templates.js]: I think, there is only one case, where server tests are suitable for ellipsis and viewport size. I think test-run-error-formatting-test is great for this purpose

File: test/functional/fixtures/regression/gh-2568/test.js (hunk 2 above)
Comment (AlexKamaev) [reply test/functional/fixtures/regression/gh-2568/test.js]: we have special tests for formatting

File: src/client/driver/command-executors/client-functions/selector-executor/filter.js [hunk 5]
```diff
@@ -4,72 +4,120 @@ import testCafeCore from '../../../deps/testcafe-core';
...
-var arrayUtils     = testCafeCore.arrayUtils;
-
-function isArrayOfNodes (obj) {
-    if (!isArray(obj))
-        return false;
-
-    for (var i = 0; i < obj.length; i++) {
-        if (!(obj[i] instanceof Node))
-            return false;
+const isArray        = Array.isArray;
```
[the remark this answers is not in the document]
Comment (miherlosev) [reply src/client/driver/command-executors/client-functions/selector-executor/filter.js]: I've create a separate issue for this -https://github.com/DevExpress/testcafe-hammerhead/issues/1721

File: src/client-functions/selectors/selector-builder.js [hunk 6]
```diff
@@ -75,22 +98,33 @@ export default class SelectorBuilder extends ClientFunctionBuilder {
     }
 
     getFunctionDependencies () {
-        var dependencies        = super.getFunctionDependencies();
-        var customDOMProperties = this.options.customDOMProperties;
-        var customMethods       = this.options.customMethods;
+        const dependencies = super.getFunctionDependencies();
+
+        const { filterVisible, filterHidden, counterMode, collectionMode, index } = this.options;
```
Comment (churkin) [inline src/client-functions/selectors/selector-builder.js]: I more like:
```javascript
const { 
    filterVisible, 
    filterHidden, 
    counterMode, 
    collectionMode, 
    index,
    customDOMProperties, 
    customMethods, 
    apiFnChain, 
    boundArgs
} = this.options;
```

File: src/client-functions/selectors/selector-builder.js [hunk 7]
```diff
@@ -75,22 +98,33 @@ export default class SelectorBuilder extends ClientFunctionBuilder {
...
+        const { customDOMProperties, customMethods, apiFnChain, boundArgs }       = this.options;
+
+        let selectorAncestor = this;
+
+        while (selectorAncestor.options.sourceSelectorBuilder)
+            selectorAncestor = selectorAncestor.options.sourceSelectorBuilder;
+
+        const apiFnID = selectorAncestor.options.apiFnID;
```
Comment (churkin) [inline src/client-functions/selectors/selector-builder.js]: extract
```javascript
let selectorAncestor = this;

while (selectorAncestor.options.sourceSelectorBuilder)
    selectorAncestor = selectorAncestor.options.sourceSelectorBuilder;

const apiFnID = selectorAncestor.options.apiFnID;
```

to the separate method

File: src/client/driver/command-executors/client-functions/selector-executor/filter.js [hunk 8]
```diff
@@ -4,72 +4,120 @@ import testCafeCore from '../../../deps/testcafe-core';
...
+
+FILTER_ERROR_TO_API_RE[SELECTOR_FILTER_ERROR.filterVisible] = /^\.filterVisible\(\)$/;
+FILTER_ERROR_TO_API_RE[SELECTOR_FILTER_ERROR.filterHidden]  = /^\.filterHidden\(\)$/;
+FILTER_ERROR_TO_API_RE[SELECTOR_FILTER_ERROR.nth]           = /^\.nth\(\d+\)$/;
+
+class SelectorFilter {
+    constructor () {
+        this.err = null;
     }
+    get error () {
```
Comment (churkin) [inline src/client/driver/command-executors/client-functions/selector-executor/filter.js]: Add new lines

File: src/client/driver/command-executors/client-functions/selector-executor/filter.js [hunk 9]
```diff
@@ -4,72 +4,120 @@ import testCafeCore from '../../../deps/testcafe-core';
...
+        return this.err;
+    }
+    set error (message) {
+        if (this.err === null)
+            this.err = message;
+    }
+    filter (node, options, apiInfo) {
+        let filtered = arrayUtils.filter(node, n => exists(n));
```
Comment (churkin) [inline src/client/driver/command-executors/client-functions/selector-executor/filter.js]: `filter(node, exists);`

File: src/client/driver/command-executors/client-functions/selector-executor/filter.js [hunk 10]
```diff
@@ -4,72 +4,120 @@ import testCafeCore from '../../../deps/testcafe-core';
...
 
-    return true;
-}
-
-function getNodeByIndex (collection, index) {
-    return index < 0 ? collection[collection.length + index] : collection[index];
-}
-
+        if (options.filterVisible) {
+            filtered = filtered.filter(n => visible(n));
```
Comment (churkin) [inline src/client/driver/command-executors/client-functions/selector-executor/filter.js]: `filtered = filtered.filter(visible);`

File: src/client/driver/command-executors/client-functions/selector-executor/filter.js [hunk 11]
```diff
@@ -4,72 +4,120 @@ import testCafeCore from '../../../deps/testcafe-core';
...
-        if (options.filterVisible)
-            filtered = filtered.filter(n => visible(n));
+        return filtered;
+    }
+    cast (node) {
+        let result = null;
 
-        if (options.filterHidden)
-            filtered = filtered.filter(n => !visible(n));
+        if (hammerhead.utils.types.isNullOrUndefined(node))
```
Comment (churkin) [inline src/client/driver/command-executors/client-functions/selector-executor/filter.js]: `const typeUtils = hammerhead.utils.types;`

File: src/client/driver/command-executors/client-functions/selector-executor/filter.js [hunk 12]
```diff
@@ -4,72 +4,120 @@ import testCafeCore from '../../../deps/testcafe-core';
...
 
-        if (options.counterMode) {
-            if (options.index !== null)
-                return getNodeByIndex(filtered, options.index) ? 1 : 0;
+        else if (node instanceof Node)
+            result = [node];
 
-            return filtered.length;
-        }
+        else if (node instanceof HTMLCollection || node instanceof NodeList || this.isArrayOfNodes(node))
```
Comment (churkin) [inline src/client/driver/command-executors/client-functions/selector-executor/filter.js]: Move to utils

File: src/client/driver/command-executors/client-functions/selector-executor/filter.js [hunk 13]
```diff
@@ -4,72 +4,120 @@ import testCafeCore from '../../../deps/testcafe-core';
...
-    },
+        return true;
+    }
+    getNodeByIndex (collection, index) {
+        return index < 0 ? collection[collection.length + index] : collection[index];
+    }
+}
+
+// Selector filter
+hammerhead.nativeMethods.objectDefineProperty.call(window, window, '%testCafeSelectorFilter%', {
```
Comment (churkin) [inline src/client/driver/command-executors/client-functions/selector-executor/filter.js]: `const nativeMethods = hammerhead.nativeMethods;`

**Comment by AlexKamaev** [comment]
please review

**Comment by AlexKamaev** [comment]
@churkin @AndreyBelym

File: src/errors/test-run/index.js [hunk 14]
```diff
@@ -70,10 +70,12 @@ export class InvalidSelectorResultError extends TestRunErrorBase {
 }
 
 export class CantObtainInfoForElementSpecifiedBySelectorError extends TestRunErrorBase {
```
Comment (AndreyBelym) [inline src/errors/test-run/index.js]: Maybe we these 3 errors from a common base class, e.g. `SelectorErrorBase`.

File: src/client-functions/selectors/add-api.js [hunk 15]
```diff
@@ -10,34 +9,33 @@ import makeRegExp from '../../utils/make-reg-exp';
 import selectorTextFilter from './selector-text-filter';
 import selectorAttributeFilter from './selector-attribute-filter';
 
-const SNAPSHOT_PROPERTIES = NODE_SNAPSHOT_PROPERTIES.concat(ELEMENT_SNAPSHOT_PROPERTIES);
+let apiFn = null;
```
Comment (AndreyBelym) [inline src/client-functions/selectors/add-api.js]: Can we avoid using the global variable for storing temporary results? You can create an internal class (e.g. `SelectorAPIBuilder`) if you have to share context between functions.

**Comment by AlexKamaev** [comment]
@testcafe-build-bot retest

**Comment by AlexKamaev** [comment]
@testcafe-build-bot retest

**Comment by AlexKamaev** [comment]
@testcafe-build-bot retest

**Pull request merged**